src/System.Private.CoreLib/src/System/Threading/ThreadPool.Unix.cs

```diff
@@ -144,22 +332,13 @@ private static IntPtr ThreadPoolDispatchCallback(IntPtr context)
              bool executeOnlyOnce,
              bool flowExecutionContext)
         {
-            //
-            // This is just a quick-and-dirty implementation to make TaskFactory.FromAsync
-            // work for the few apps that are using it.  A proper implementation would coalesce
-            // multiple waits onto a single thread, so that fewer machine resources would be
-            // consumed.
-            //
-
-            Debug.Assert(executeOnlyOnce);
-
-            QueueUserWorkItem(_ =>
-            {
-                bool timedOut = waitObject.WaitOne((int)millisecondsTimeOutInterval);
-                callBack(state, timedOut);
-            });
-
-            return new RegisteredWaitHandle();
+            RegisteredWaitHandle registeredHandle = new RegisteredWaitHandle(
+                waitObject,
+                new _ThreadPoolWaitOrTimerCallback(callBack, state, flowExecutionContext),
+                (int)millisecondsTimeOutInterval,
+                !executeOnlyOnce);
+            ClrThreadPool.ThreadPoolInstance.RegisterWaitHandle(registeredHandle);
```

Comment:
After adding the finalizer to RegisteredWaitHandle, need to consider what would happen if this throws OOM. The RegisteredWaitHandle somehow needs to be told that it does not need to unregister itself in the finalizer since the wait was not actually registered. Maybe something like:

```c#
bool registered = false;
try
{
    ClrThreadPool.ThreadPoolInstance.RegisterWaitHandle(registeredHandle);
    registered = true;
}
finally
{
    if(!registered)
        registeredHandle.OnRegistrationFailed(); // or something, which could call GC.SuppressFinalize(this)
}
```